### PR TITLE
More comprehensive WorkerState task counters

### LINKED ISSUE #6319: Inconsistent Worker.waiting_for_data_count
At the end of these tests, Worker.waiting_for_data_count does not match the number of tasks where ts.waiting_for_data is true:
```
FAILED distributed/tests/test_client.py::test_persist_workers_annotate - Asse...
FAILED distributed/tests/test_client.py::test_persist_workers - AssertionError
FAILED distributed/tests/test_failed_workers.py::test_submit_after_failed_worker_async
FAILED distributed/tests/test_failed_workers.py::test_worker_who_has_clears_after_failed_connection
FAILED distributed/tests/test_steal.py::test_steal_communication_heavy_tasks
FAILED distributed/tests/test_worker.py::test_gather_dep_exception_one_task_2
FAILED distributed/tests/test_worker.py::test_missing_released_zombie_tasks
FAILED distributed/tests/test_worker.py::test_missing_released_zombie_tasks_2
FAILED distributed/tests/test_worker.py::test_task_flight_compute_oserror - A...
FAILED distributed/tests/test_cancelled_state.py::test_worker_find_missing - ...
FAILED distributed/tests/test_cancelled_state.py::test_worker_stream_died_during_comm
FAILED distributed/tests/test_cancelled_state.py::test_flight_to_executing_via_cancelled_resumed
FAILED distributed/tests/test_cancelled_state.py::test_flight_cancelled_error
FAILED distributed/tests/test_cancelled_state.py::test_in_flight_lost_after_resumed
FAILED distributed/tests/test_worker_state_machine.py::test_fetch_to_compute
```

#6318 temporarily disables this check in ``Worker.validate_state()``.

### PR DESCRIPTION
**Pull request by crusaderky**
- Expose a more comprehensive and accurate view of the number of tasks in each state on the worker to the scheduler (through the heartbeat) and to prometheus. Crucially, this PR gives us a measure of network saturation and co-assignment by showing how many tasks are queued up in `fetch` state.
- O(1) count number of tasks in fetch state
- Closes #6319
- Disallow directly writing to Worker.data. I am quite frankly baffled this was possible before. I don't think we should have a deprecation cycle for this?
- Prometheus tasks metric `stored` has been renamed to `memory` CC @ntabris

### CODE REVIEW

File: distributed/http/worker/prometheus/core.py [hunk 1]
```diff
@@ -26,15 +28,15 @@ def __init__(self, server: Worker):
...
             self.build_name("tasks"),
             "Number of tasks at worker.",
             labels=["state"],
         )
-        tasks.add_metric(["stored"], len(self.server.data))
-        tasks.add_metric(["executing"], self.server.state.executing_count)
-        tasks.add_metric(["ready"], len(self.server.state.ready))
-        tasks.add_metric(["waiting"], self.server.state.waiting_for_data_count)
+        for k, n in ws.task_counts.items():
+            tasks.add_metric([k], n)
```
Comment (crusaderky) [inline distributed/http/worker/prometheus/core.py:39]: Renamed stored -> memory

File: distributed/worker_state_machine.py [hunk 2]
```diff
@@ -3244,6 +3264,33 @@ def _to_dict(self, *, exclude: Container[str] = ()) -> dict:
         info = {k: v for k, v in info.items() if k not in exclude}
         return recursive_to_dict(info, exclude=exclude)
 
+    @property
+    def task_counts(self) -> dict[TaskStateState | Literal["other"], int]:
+        # Actors can be in any state other than {fetch, flight, missing}
+        n_actors_in_memory = sum(
+            self.tasks[key].state == "memory" for key in self.actors
+        )
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:3272]: This is O(n) but actors are such a niche feature that we should not care in 99% of the cases

File: distributed/http/worker/prometheus/core.py (hunk 1 above)
Comment (ntabris) [reply distributed/http/worker/prometheus/core.py:39]: what's the difference between things in `data` and actors in "memory" state? (I noticed that actors weren't included in previous metric and are included in new metric)

Comment (crusaderky) [reply distributed/http/worker/prometheus/core.py:39]: The "real" instance of an actor lives in WorkerState.actors instead of data. Its proxies on remote workers live in data. Actors are an edge case I would not spend time fine-tuning anything towards anyway.

**Comment by crusaderky** [comment]
Ready for review and merge. All test failures seem unrelated.

File: distributed/worker_state_machine.py [hunk 3]
```diff
@@ -3244,6 +3264,33 @@ def _to_dict(self, *, exclude: Container[str] = ()) -> dict:
...
+            "ready": len(self.ready),
+            "constrained": len(self.constrained),
+            "waiting": len(self.waiting),
+            "fetch": self.fetch_count,
+            "missing": len(self.missing_dep_flight),
+            # Also includes cancelled(flight) and resumed(flight->waiting)
+            "flight": len(self.in_flight_tasks),
+        }
+        # released | error
+        out["other"] = other = len(self.tasks) - sum(out.values())
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/worker_state_machine.py:3290]: It felt a bit useless, from the point of view of performance metrics. An erred task is just sitting there until it's either released or its exception is retrieved.

[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/worker_state_machine.py:3290]: Fair point, I guess a counter of tasks transitioning into this state would be the interesting bit here rather than any point-in-time count.
